docs/protocol-initialization-facet.md

```diff
@@ -0,0 +1,57 @@
+
+Two biggest benefits of the Protocol initialization handler are:
+- It allows atomic cuts of multiple facets, so there is generally no need to pause the protocol during the upgrade.
+- It allows custom storage manipulation, so if any back filling is needed, we have a way to do it.
+
+### Versioning
+
+Initialize accepts version encoded in bytes32. Every time upgrade takes place, an unique version must be supplied, otherwise initialization reverts. Depending on version, some additional checks can be done, for example upgrade to certain versions can only be done if current protocol version is exactly one version lower. This kind of restriction is planned to be used on every upgrade. This means that skipping versions during the upgrade will not be possible. If the latest version of protocol code is more than one version higher that current deployed protocol, all intermediate upgrades need to be done to upgrade to the lates version.
+
+Facets allows to query the current version by calling `getVersion`.
+
+### Facet initialization
+
+Each Protocol facet should have initialization function (even if no-op) to ensure consistency across codebase and different version. Initialization function should be written as if protocol is deployed for the first time (i.e. not taking into account potential effects on the upgrade).
+
+When protocol is deployed for the first time, `initialize` on Protocol initialization handler should get list of all facet implementation addresses together with corresponding initialization data. This data is passed directly on individual facets where whole initialization takes place.  
+
+However, before upgrade is done, effects of facet initializers should carefully be considered before passing data to initialization handler. If facet's initialization function does not affect the protocol state in a harmful way, it should be passed to initialization handler in the same way as for the initial deployment. However, when initialize affects the storage (for example Config handler sets all counters to 1), facet should be omitted from initialization call and approach from next section should be followed.
+
+### Data initialization
+
+When facet initializers are clashing with existing data or simply a specific storage must be populated, initializer facet must be updated to cater for these kind of change. Suppose new version is X.Y.Z and there is a need to modify the storage during the update, a new internal function should be prepared, called `initVX_Y_Z`. This function can make additional version checks and can accept arbitrary data which can be handled in any desired way. This allows to populate custom storage slot during the upgrade or mimic actions of the facet initializers that would otherwise be harmful. For example if another limit is added to the Config handler, `initVX_Y_Z` could simply store new value to desired location, without the need to overwrite other config values and at the same time leaves counters intact.  
```

Comment:
```suggestion
When facet initializers clash with existing data or simply a specific storage must be populated, initializer facet must be updated to cater for this kind of change. Suppose new version is X.Y.Z and there is a need to modify the storage during the update, a new internal function should be prepared, called `initVX_Y_Z`. This function can make additional version checks and can accept arbitrary data which can be handled in any desired way. This allows populating custom storage slots during the upgrade or mimicking actions of the facet initializers that would otherwise be harmful. For example if another limit is added to the Config handler, `initVX_Y_Z` could simply store new value to desired location, without the need to overwrite other config values and at the same time leaves counters intact.  
```